docs/proposals/resource_management.md

```diff
@@ -0,0 +1,381 @@
+
+Pod5 is classified as Best effort and is hence nested under the BE cgroup  
+
+/BE/cpu.shares = 2  
+/BE/cpu.quota= not set  
+/BE/m.limit_in_bytes = Allocatable - 7Gi  
+/BE/Pod5/m.limit = no limit  
+
+Note:
+- Best Effort pods are treated as lowest priority. So for the BE QoS cgroup we set cpu shares to the lowest possible value ie.2. This ensures that the best effort containers get a relatively small share of cpu time.
+- Also we don't set any quota on the cpu resources as the containers on the BE pods can use any amount of free resources on the node.
+- Having memory limit of BE cgroup as (Allocatable - summation of memory limits of Guaranteed and Burstable pods) would result in BE pods becoming more susceptible to being OOM killed. As more G and Bu pods are scheduled kubelet will more likely kill BE pods, even if the G and Bu pods are using less than their request since we will be dynamically reducing the size of BE m.limit_in_bytes. But this allows for better memory guarantees to the G and Bu pods.
+
+### With Systemd
+
+In systemd we have slices for the three top level QoS class. Further each pod is a subslice of exactly one of the three QoS slices. Each container in a pod belongs to a scope nested under the qosclass-pod slice.  
+
+Example:  
+We plan to have the following cgroup hierarchy on systemd systems
+
+```
+/memory/G-PodUID.slice/containerUID.scope
+/cpu,cpuacct/G-PodUID.slice/containerUID.scope
+/memory/Bu.slice/Bu-PodUID.slice/containerUID.scope
+/cpu,cpuacct/Bu.slice/Bu-PodUID.slice/containerUID.scope
+/memory/BE.slice/BE-PodUID.slice/containerUID.scope
+/cpu,cpuacct/BE.slice/BE-PodUID.slice/containerUID.scope
+```
+
+There are similar slices for Burstable and Best Effort QoS classes.
+
+## Implementation Plan
+
+The implementation plan is outlined in the next sections.
+We will have a 'enable-qos-cgroups' flag to specify if the user wants to use the QoS based cgroup hierarchy. The flag would be set to false by default atleast in v1.4.
+
+#### Top level Cgroups for QoS tiers
+
+Two top level cgroups for Burstable and Best effort Qos classes are created when Kubelet starts to run on a node. All guaranteed pods cgroups are by default nested under the `Root`. So we dont create a top level cgroup for the Guaranteed class. For raw cgroup systems we would use libcontainers cgroups manager for general cgroup management(cgroup creation/destruction). But for systemd we don't have equivalent support for slice management in libcontainer yet. So we will be adding support for the same in the Kubelet. These cgroups are only created once on Kubelet initialization as a part of node setup.  
+
+`Note:  
+The Guaranteed pods cgroup would be nested directly under the root by default.
+But we will have a "enable-guaranteed-cgroup" flag that would allow for having a specific Guaranteed Cgroup under the root.`
+
+#### Pod level Cgroup creation and deletion (Docker runtime)
+
+- When a new pod is brought up, its QoS class is firstly determined.
```

Comment:
Re: automatic QoS classification. Making the API backwards-compatible is a great goal, and auto classification seems like a good way to go. However, as a user submitting critical pods it would be nice to guarantee the right isolation strategy is going to be applied without replicating the classification logic on the client side. What do you think about adding an optional explicit QoS class field to the pod spec? When absent, it could be filled in on the user's behalf. When present it could be used to validate the resource shape conforms to the intended class.